source/brailleDisplayDrivers/albatross.py

```diff
@@ -0,0 +1,437 @@
+		# in single ctrl-key presses and key combinations the first key is resent as last one.
+		if ord(data) in CONTROL_KEY_CODES:
+			# at most 4 keys.
+			try:
+				data += self._dev.read(4)
+			except serial.serialutil.SerialException:
+				log.debug("Ctrl key read failed")
+				return
+		pressedKeys = set(data)
+		try:
+			inputCore.manager.executeGesture(InputGestureKeys(pressedKeys))
+		except (inputCore.NoInputGestureAction, AttributeError):
+			pass
+
+	def _checkConnection(self, data: bytes) -> bool:
+		# If numCells > 0, there is working connection.
+		if self.numCells:
+			return True
+		# If no connection, Albatross sends continuously byte \xff
+		# followed by byte containing various settings like number of cells.
+		# But there may be garbage before \xff.
+		if data != INIT_START_BYTE:
+			try:
+				while data != INIT_START_BYTE and len(data):
+					data = self._dev.read(1)
+			except serial.serialutil.SerialException:
+				log.debug("Init byte read failed")
+				return False
+		if not len(data):
+			log.debug("No init byte")
+			return False
+		try:
+			data = self._dev.read(1)
+		except serial.serialutil.SerialException:
+			log.debug("Value byte read failed")
+			return False
+		if not len(data):
+			log.debug("No value byte")
+			return False
+		log.debug("Value byte: %r" % data)
+		if not self._sendToDisplay(ESTABLISHED):
+			log.debug("Connection establishment failed")
+			return False
+		# If bit 7 (LSB 0 scheme) is 1, there is 80 cells model, else 46 cells model.
+		# Other display settings are currently ignored so skipping separate function
+		# definition this time.
+		self.numCells = 80 if ord(data) >> 7 == 1 else 46
+		# Reconnected if length of _oldCells is numCells. Show last known content.
+		if len(self._oldCells) == self.numCells:
+			self.display(self._currentCells)
+		return True
+
+	def _exitInternalMenu(self, data: bytes) -> bool:
+		# It is possible that there is no connection from perspective of display.
+		# If user has entered to the internal menu of device, display starts to send
+		# initial packet containing \xff and settings byte when user exits the menu.
+		if data != INIT_START_BYTE:
+			return True
+		# Other display settings are currently ignored, and number of cells is already known.
+		# Only log value byte for debugging.
+		try:
+			data = self._dev.read(1)
+		except serial.serialutil.SerialException:
+			log.debug("Value byte read after internal menu exit failed")
+			return False
+		if not len(data):
+			log.debug("No value byte after internal menu exit")
+		else:
+			log.debug("Value byte after internal menu exit: %r" % data)
+		if not self._sendToDisplay(ESTABLISHED):
+			log.debug("Connection establishment failed")
+			return False
+		self._clearOldCells()
+		# Using _currentCells to show last known content.
+		self.display(self._currentCells)
+		return True
+
+	def display(self, cells: List[int]):
+		# Keep _currentCells up to date.
+		self._currentCells = cells.copy()

```

Comment:
For performance, you might consider caching the calculated data (output) rather than input.